### PR TITLE
Initial attempt on an elytra pathfinder.

### PR DESCRIPTION
**Pull request by FriedrichM**
I integrated my previously standalone Elytra Pathfinder as a new function of AutoWalk.
This might not be perfect but it provides a base upon which improvements can be made.

### CODE REVIEW

**Comment by FriedrichM** [comment]
oh  btw here is a how to use: turn on efly, look in the direction you want to go (straights and diagonals are supported) and activate autowalk in pathfinder mode

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt [hunk 1]
```diff
@@ -26,7 +32,9 @@ import net.minecraftforge.fml.common.network.FMLNetworkEvent
 )
 class AutoWalk : Module() {
     @JvmField
-    var mode: Setting<AutoWalkMode> = register(Settings.e("Direction", AutoWalkMode.BARITONE))
+    var mode: Setting<AutoWalkMode> = register(Settings.e("Direction", AutoWalkMode.PATHFIND))
+    val flydelay = register(Settings.integerBuilder("SableFlightDelay").withRange(0, 400).withValue(0).withVisibility { mode.value == AutoWalkMode.PATHFIND }.build())
+    private val drawPath = register(Settings.booleanBuilder("DrawPath").withValue(false).withVisibility { mode.value == AutoWalkMode.PATHFIND }.build())
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt:37]: 1. shouldn't this be on by default?
2. are you still using the bleh GL code or did you switch to Baritone's?

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt [hunk 2]
```diff
@@ -41,28 +49,30 @@ class AutoWalk : Module() {
...
 
         if (mc.player == null) {
             disable()
             return
         }
 
-        if (mc.player.isElytraFlying) {
-            sendErrorMessage("$chatName Baritone mode isn't currently compatible with Elytra flying! Choose a different mode if you want to use AutoWalk while Elytra flying")
-            disable()
+        if (MODULE_MANAGER.isModuleEnabled(ElytraFlight::class.java)) {
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt:73]: There should probably be an option called "ElytraFlightCheck", so you don't need to use KAMI Blue's elytraflight. It is the best but I like stuff being as customizable as possible.

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt [hunk 3]
```diff
@@ -81,11 +91,14 @@ class AutoWalk : Module() {
...
+            "ELYTRA"
+        } else if (BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.goal != null) {
             direction!!
         } else {
             when (mode.value) {
-                AutoWalkMode.BARITONE -> "NONE"
+                AutoWalkMode.PATHFIND -> "NONE"
                 AutoWalkMode.FORWARD -> "FORWARD"
                 AutoWalkMode.BACKWARDS -> "BACKWARDS"
             }
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt:104]: This shouldn't be changed, it should still return the direction it's going in.

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt [hunk 4]
```diff
@@ -113,11 +126,23 @@ class AutoWalk : Module() {
...
+        if (drawPath.value && mode.value.equals(AutoWalkMode.PATHFIND) && MODULE_MANAGER.isModuleEnabled(ElytraFlight::class.java) && Astarpathfinder.path != null) {
+            val lines: MutableList<Vec3d> = mutableListOf()
+            for (i in 0 until Astarpathfinder.path.size) {
+                lines.add(Vec3d(Astarpathfinder.path[i].pos.x + 0.5, mc.player.posY - 0.5, Astarpathfinder.path[i].pos.z + 0.5))
+            }
+            GlStateManager.enableTexture2D()
+            KamiTessellator.drawLinesFromList(lines, 1f, 1f, 0f, 1f)
+            KamiTessellator.releaseGL()
+        }
+    }
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt:147]: I'll probably rewrite this GL myself before merging

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/ElytraFlight.kt [hunk 5]
```diff
@@ -501,18 +522,39 @@ class ElytraFlight : Module() {
...
+        if (oldpathfind == false && pathfind == true) {
+            effectiveMode = ElytraFlightMode.CONTROL
+            Astarpathfinder.init()
+        }
+        if (pathfind == false) {
+            effectiveMode = mode.value
+            if (oldpathfind == true)
+                Astarpathfinder.turnOff()
+        }
+
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/movement/ElytraFlight.kt:536]: Should this really be running when the game is still loading?

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt (hunk 4 above)
Comment (FriedrichM) [reply src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoWalk.kt:147]: feel free :)

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/ElytraFlight.kt (hunk 5 above)
Comment (FriedrichM) [reply src/main/java/me/zeroeightsix/kami/module/modules/movement/ElytraFlight.kt:536]: I could solve this by using a settingslistener on the mode setting from AutoWalk and call some "updatePathfindState" function from autowalk.onenable.  ill push my idea you can have a look at the onenable/disable functions of both modules

**Comment by FriedrichM** [comment]
i have noticed a bug where it would sometimes get into standing pose, get stuck in the block above the player and take damage when its landing because it could find a path through an obstacle.  i didnt find the cause yet, you can have a look  yourself if you want. just fly 1 block below the celing in a path that is blocked after a while

**Comment by dominikaaaa** [comment]
That happened when switching modes in elytraflight, so that's why the setting listener resets that when the setting is changed

**Comment by FriedrichM** [comment]
ill have a look if its related to something like that tomorrow or the day after

File: src/main/java/me/zeroeightsix/kami/util/astaralg/Astarpathfinder.java [hunk 6]
```diff
@@ -219,7 +220,7 @@ private static void pathfind() {
                 stuckPos = getPlayerBlockpos();
                 state = State.LAND_SAVE;
                 nopathfound = 0;
-                currentEvent.cancel();
+                elytraFlight.controlMode(currentEvent);
```
Comment (FriedrichM) [inline src/main/java/me/zeroeightsix/kami/util/astaralg/Astarpathfinder.java:223]: i think this was the problem , i canceledone packet before starting the autoland

Comment (FriedrichM) [reply src/main/java/me/zeroeightsix/kami/util/astaralg/Astarpathfinder.java:223]: that didn't fix it completely. still happens sometimes if you fly directly under a block but a lot less often but still. I am thinking about making a detection system for this issue

**Comment by dominikaaaa** [comment]
Something that I noticed which is a pretty big flaw, it doesn't know how to go up. Control mode can go up by "pressing" space in default settings on all servers, so it should at least be optional

**Comment by FriedrichM** [comment]
upwards and downwards flight is a tough thing because you cant go straight up, this means i would have to 
check if there is enough space ahead to go up which is hard to implement in a graph based path finding algorithm.
Atm every block in my current y level is a node and it searches for a short way through the resulting graph.
My best idea to add upwards flight is to add edges in the graph from the each node to the node form the y level above that you'd reach if you start flying up from that node. (only add this edge if there is nothing in the way ofc.) In general this is a lot of work and i don't see a huge benefit for highway travel as its mostly 3/4 levels. I could add the option to press space/shift  to manually change y level during flight. Although i prob would postpone this feature to the first update of the module.

**Pull request closed without merging**